### DISCUSSION TITLE
Blazor CRUD UI

### DISCUSSION THREAD

**Discussion by matt-goldman** [pdevito3/craftsman#67] [Ideas]
Hi @pdevito3 

Just putting this out there for discussion, I know this would be a huge undertaking. But what do you think about the idea of generating (or scaffolding) a basic CRUD UI in Blazor? Wouldn't need to be anything fancy, just using the basic out of the box styles, a form for adding/editing each entity and a table for listing them.

What do you think?

**Comment by pdevito3** [comment]
Hey! 

So I have had a lot of requests for this and it's been floating on my radar for a while. 

The big downside here is... I've seen some of Steve's presentations, but that's about as far as my blazor xp goes 🙃, so I'd be hesitant to put something out of likely lower quality. Honestly, MS's history with UI frameworks isn't inspiring either but that's a different issue. 

FWIW, I am putting out a first pass at a Duende BFF with react this next release to provide a front end base, so that could provide a foundation for adding a blazor option, but idk if I'd get to it immediately. With that said, I might be open to a community contribution for it if we were able to come up with a good plan. My only hesitation there is the maintenance overhead that would come after the fact.

I will probably be completely overhauling the auth server implementation the following release, so that could be an opportunity to try blazor instead of MVC/Razor too. 

So short answer, maybe in the near term if it's community driven. If not, maybe I'd do it solo in the future. 

Thoughts?

**Comment by matt-goldman** [comment]
Great to hear it's on your radar I'd share exactly the same concerns to be honest. I've got too many side projects at the moment but if I get some time I'll take a swing at it :)

**Comment by Jogai** [comment]
Another consideration is the audience. With a react (or vue/svelte or any other js framework) frontend you might attract frontend developers in search for a backend. That audience is bigger than blazor devs. With blazor it will look more like a .net only solution. Both are fine, but it depends a bit on what the project's goals are.

**Comment by thestacks** [comment]
This is a great idea! I think that Blazor is a really stable solution and it seems that it won't be 'killed' that fast looking at its community contribution and its potential. 😄 
Personally I love using Blazor, and I can help you with that as I have plenty of time after hours! 🔥

🧠 What I think would be good here is to provide a MVP with Blazor at first. Then (as community can contribute to this), leave people opportunity to pick whether they want Blazor, React or any other framework to be used as a frontend. So you would use craftsman as usual, but when attaching frontend, you get to pick framework.

**Reply by pdevito3** [reply]
if you (and or others) are able to put together an example project with rules around different setup paths, then let's talk about it

**Comment by thestacks** [comment]
🆕 🆕 🆕 
I have created simple CRUD application with full state management. This is rather PoC (or sample project), but it's fully functional, responsive and... (I think) nice looking.
This does not have any craftsman integration yet - I just want to know if this is the direction you want to go in.

👀  Check it out and read the README.md file (it contains some screenshots and explanation): https://github.com/thestacks/craftsman-sample-blazor

**Reply by pdevito3** [reply]
Sweet! I'll take a look in the next few days. 

Thanks! 🍻

**Reply by pdevito3** [reply]
hey, got to take a look at this a bit tonight. thanks again for putting some time into this!

generally, it looks solid, but a topics of discussion for me:

- i have mixed feelings about the global store for state management. i know everyone has their own flavor of state management and different shops have different styles, but just using a global store like this has seen downsides in the react ecosystem that I could see coming up here too. curious to see if blazor has a phase for this like react did.
- components are always an interesting topic for me. there's always a fun line of not reinventing the wheel, but also not having to be boxed in with a super opinionated framework. 
- that leads to (another) opinionated topic of css  and styling. i'm a huge tailwind fan and would like to promote that in the framework, but again, i know not everyone is on that train and some people will want different options here.

with all that said, some opinions need to be laid down as a base to keep the scaffolding at any type on maintainability so this is probably all more of a discussion topic for the blazor focused group to see where everyone lands. i'm sure a good discussion could be had here with a happy landing zone for the community.

as far as integrating it into craftsman, i am still worried about the amount of overhead this would bring even if we had solid answers for everything else. add on top of that the cognitive overhead of the blazor side of things i can take a look at and understand, but actively supporting it and pushing forward strong recommendations as questions come up and rubber meets the road would be tough right now.

part of the goal of this project was to give a good base and for people to be able to easily fork it and modify it to a style that fits their shops, so i always want that to be a pillar of how people can use this. i haven't fully gotten the craftsman POC refactored to be super amenable to that, but there are some people using it like that already. 

i'd lean toward having this scaffolding be a separate community supported project that I'd be happy to promote in the docs and within the wrapt community, but i don't think i have the bandwidth to support this as well right now. maybe that'll change in the future and we can merge it in, but i don't think that's the case at the moment. happy discuss any work around it, just can't support it myself is all

regardless, i appreciate you putting the time into this!

**Comment by pdevito3** [comment]
Since I have a few Blazor enthusiasts on this thread: 

I'm working on a new Auth Server template which I'm thinking might be a good option to use with Blazor. If anyone's interested in hopping in on it, let me know!
